File: 2021/a.hs
#!/usr/bin/env runghc

-- ACM-ICPC 2021 Domestic Qualifier
-- Problem A: Marbles Tell Your Lucky Number

import Control.Monad (when)
import Control.Applicative ((<$>))
import Data.List (sort)

main :: IO ()
main = do
    cups <- map read . words <$> getLine :: IO [Int]
    when (cups /= [0, 0, 0, 0]) $ do
        print $ lucky cups
        main

lucky :: [Int] -> Int
lucky cups = lucky' . filter ((>0) . fst) $ zip cups [0..]

lucky' :: [(Int, Int)] -> Int
lucky' cs =
    let ((m, i) : cs') = sort cs in
        if null cs' then m
        else lucky' $ (m, i) : [(m' - m, j) | (m', j) <- cs', m' > m]

File: 2021/a.ex
#!/usr/bin/env elixir

# ACM-ICPC 2021 Domestic Qualifier
# Problem A - Marbles Tell Your Lucky Number

defmodule Main do
  def main do
    cups = IO.gets("") |> String.split |> Enum.map(&String.to_integer/1)
    if cups != [0, 0, 0, 0] do
        solve(cups) |> IO.inspect
        main()
    end
  end

  def solve(cups) do
     cups |> Enum.filter(&(&1 > 0)) |> lucky
  end

  def lucky([m]), do: m
  def lucky(cs) do
    [m | cs] = Enum.sort(cs)
    lucky([m | (for m2 <- cs, m2 > m, do: m2 - m)])
  end
end

Main.main
